File: 백준/Gold/10026. 적록색약/적록색약.cc
#include<bits/stdc++.h>
using namespace std;
int n;
string s;
char normal[104][104];
char special[104][104];
int visited[104][104];
// DFS => connected components
int dx[4] = { -1,1,0,0 };
int dy[4] = { 0,0,-1,1 };
int cnt = 0;
void go(int y, int x, int type) {
	visited[y][x] = 1;
	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i]; int nx = x + dx[i];
		if (nx < 0 || ny < 0 || nx >= n || ny >= n || visited[ny][nx]) continue;
		if (type && special[y][x] == special[ny][nx]) go(ny, nx, type);
		else if (normal[y][x] == normal[ny][nx]) go(ny, nx, type);
	}
	return;
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> s;
		for (int j = 0; j < n; j++) {
			normal[i][j] = s[j];
			if (s[j] == 'R' || s[j] == 'G') special[i][j] = 'R';
			else special[i][j] = 'B';
		}
	}

	// normal
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (!visited[i][j]) {
				go(i, j, 0); cnt++;
			}
			
		}
	}
	cout << cnt << " ";
	cnt = 0; fill(&visited[0][0], &visited[0][0] + 104 * 104, 0);

	// special
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (!visited[i][j]) {
				go(i, j, 1); cnt++;
			}

		}
	}
	cout << cnt;

	return 0;
}

File: 백준/Gold/10026. 적록색약/적록색약.py
import sys
sys.setrecursionlimit(10000)

n = int(input())
normal = []
special = []
visited = [[0] * 104 for _ in range(104)]
cnt = 0

# 입력 및 normal/special 분리 처리
for _ in range(n):
    row = input()
    normal.append(list(row))
    special.append(['R' if c in ['R', 'G'] else 'B' for c in row])

# 상하좌우 방향 벡터
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

# DFS 탐색 함수
def dfs(y, x, type_):
    visited[y][x] = 1
    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]
        if 0 <= ny < n and 0 <= nx < n and not visited[ny][nx]:
            if type_ == 1 and special[y][x] == special[ny][nx]:
                dfs(ny, nx, type_)
            elif type_ == 0 and normal[y][x] == normal[ny][nx]:
                dfs(ny, nx, type_)

# 일반인 시각
for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            dfs(i, j, 0)
            cnt += 1
print(cnt, end=' ')

# 방문 배열 초기화
cnt = 0
visited = [[0] * 104 for _ in range(104)]

# 적록색약 시각
for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            dfs(i, j, 1)
            cnt += 1
print(cnt)
